src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/MetadataTests/DefaultJsonPropertyInfo.cs

```diff
@@ -0,0 +1,411 @@
+
+            Assert.Equal(nameof(ClassWithRequiredCustomAttributes.NonRequired), typeInfo.Properties[0].Name);
+            Assert.False(typeInfo.Properties[0].IsRequired);
+
+            Assert.Equal(nameof(ClassWithRequiredCustomAttributes.RequiredA), typeInfo.Properties[1].Name);
+            Assert.True(typeInfo.Properties[1].IsRequired);
+
+            Assert.Equal(nameof(ClassWithRequiredCustomAttributes.RequiredB), typeInfo.Properties[2].Name);
+            Assert.True(typeInfo.Properties[2].IsRequired);
+
+            ClassWithRequiredCustomAttributes obj = new();
+            string json = """{"NonRequired":null,"RequiredA":null,"RequiredB":null}""";
+            Assert.Equal(json, JsonSerializer.Serialize(obj, options));
+
+            var deserialized = JsonSerializer.Deserialize<ClassWithRequiredCustomAttributes>(json, options);
+            Assert.Null(deserialized.NonRequired);
+            Assert.Null(deserialized.RequiredA);
+            Assert.Null(deserialized.RequiredB);
+
+            JsonException exception = Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<ClassWithRequiredCustomAttributes>("{}", options));
+            Assert.DoesNotContain(nameof(ClassWithRequiredCustomAttributes.NonRequired), exception.Message);
+            Assert.Contains(nameof(ClassWithRequiredCustomAttributes.RequiredA), exception.Message);
+            Assert.Contains(nameof(ClassWithRequiredCustomAttributes.RequiredB), exception.Message);
+
+            exception = Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<ClassWithRequiredCustomAttributes>("""{"NonRequired":"foo"}""", options));
+            Assert.DoesNotContain(nameof(ClassWithRequiredCustomAttributes.NonRequired), exception.Message);
+            Assert.Contains(nameof(ClassWithRequiredCustomAttributes.RequiredA), exception.Message);
+            Assert.Contains(nameof(ClassWithRequiredCustomAttributes.RequiredB), exception.Message);
+
+
+            exception = Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<ClassWithRequiredCustomAttributes>("""{"NonRequired":"foo", "RequiredB":"bar"}""", options));
+            Assert.DoesNotContain(nameof(ClassWithRequiredCustomAttributes.NonRequired), exception.Message);
+            Assert.Contains(nameof(ClassWithRequiredCustomAttributes.RequiredA), exception.Message);
+            Assert.DoesNotContain(nameof(ClassWithRequiredCustomAttributes.RequiredB), exception.Message);
+
+            exception = Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<ClassWithRequiredCustomAttributes>("""{"NonRequired":"foo", "RequiredA":null}""", options));
+            Assert.DoesNotContain(nameof(ClassWithRequiredCustomAttributes.NonRequired), exception.Message);
+            Assert.DoesNotContain(nameof(ClassWithRequiredCustomAttributes.RequiredA), exception.Message);
+            Assert.Contains(nameof(ClassWithRequiredCustomAttributes.RequiredB), exception.Message);
+        }
+
+        [ConditionalFact]
```

Comment:
some note here: all `[ConditionalFact]` may throw `SkipTestException` which will show them as yellow in UI and skipped in the console. Basically I use them for tests which modify contract since one of the variants of the tests runs unwrapped JsonSerializerContext and in order for that to be unwrapped (unwrapped = used directly as TypeInfoResolver) I cannot use modifiers which means that for that variant they will be skipped